lib/policies/redirectPolicy.ts

```diff
@@ -44,23 +65,40 @@ function handleRedirect(
   const locationHeader = response.headers.get("location");
   if (
     locationHeader &&
-    (status === 300 || status === 307 || (status === 303 && request.method === "POST")) &&
-    (!policy.maxRetries || currentRetries < policy.maxRetries)
+    (status === 300 ||
+      (status === 301 && ["GET", "HEAD"].includes(request.method)) ||
+      (status === 302 && ["GET", "POST", "HEAD"].includes(request.method)) ||
+      (status === 303 && "POST" === request.method) ||
+      status === 307) &&
+    ((request.redirectLimit !== undefined && currentRetries < request.redirectLimit) ||
+      (request.redirectLimit === undefined && currentRetries < policy.maxRetries))
   ) {
     const builder = URLBuilder.parse(request.url);
     builder.setPath(locationHeader);
     request.url = builder.toString();
 
-    // POST request with Status code 303 should be converted into a
+    // POST request with Status code 302 and 303 should be converted into a
     // redirected GET request if the redirect url is present in the location header
-    if (status === 303) {
+    // reference: https://tools.ietf.org/html/rfc7231#page-57 && https://fetch.spec.whatwg.org/#http-redirect-fetch
+    if ((status === 302 || status === 303) && request.method === "POST") {
       request.method = "GET";
+      delete request.body;
     }
 
     return policy._nextPolicy
       .sendRequest(request)
-      .then((res) => handleRedirect(policy, res, currentRetries + 1));
+      .then((res) => handleRedirect(policy, res, currentRetries + 1))
+      .then((res) => recordRedirect(res, request.url));
   }
 
   return Promise.resolve(response);
 }
+
+function recordRedirect(response: HttpOperationResponse, redirect: string): HttpOperationResponse {
```

Comment:
I spent some time thinking about this, since it's somewhat non-obvious that this methods executes in the context of the penultimate call to `handleRedirect` and thus has the final url, but I can't think of a better way to do it that isn't more code.

Partly I wish we had written this method using async/await and used a while loop to do following instead of recursion, but I won't force that much of a rewrite upon you. 😄

Reply:
Add a few words to the comment to help clarify.